neo/public/javascripts/cache/services.js

```diff
@@ -0,0 +1,356 @@
+           maxY : bbox[3]
+       }
+
+
+       if(cachedRegion != undefined && turf.difference(currentRequestPolygon,cachedRegion) == undefined) {
+              //cache HIT
+//              var t0 = performance.now();
+             var result = cachedCityPolygonTree.search(item);
+             data_response = turf.featureCollection(result);
+             deferred.resolve(data_response);
+//              console.log("Cache HIT took",performance.now()-t0);
+             return deferred.promise();
+
+       }else{
+                   //cache MISS
+//               var MISSt0 = performance.now();
+              var buffered = turf.buffer(currentRequestPolygon, preFetchDistance , 'miles');
+              var bboxBuffer = turf.bbox(buffered);
+                   //Pre Fetch
+              var rteExtends = "city/" + bboxBuffer[3] + "/" + bboxBuffer[1] + "/" + bboxBuffer[2] + "/" + bboxBuffer[0];
+
+              $http.get(rteExtends).success(function(data) {
+
+                      var result_set = insertIntoTree(data.features,currentRequestPolygon).done(function(){
+                               console.log("done");
+                               var extendPoly = turf.bboxPolygon(bboxBuffer);
+
+                               if(cachedRegion == undefined)
+                                   {cachedRegion = extendPoly;}
+                               else
+                                   {cachedRegion = turf.union(extendPoly,cachedRegion);}
+
+                               var result = cachedCityPolygonTree.search(item);
+                               data_response = turf.featureCollection(result);
+
+                               deferred.resolve(data_response);
+
+                      });
+              }).error(function(data) {
+                                    console.error("Load city data failure");
+              });
+//               console.log("Cache MISS took",performance.now()-MISSt0);
+              return deferred.promise();
+       }
+  }
+
+ var insertIntoTree = function insertIntoTree(features,currentRequest){
+
+       var deferred = new $.Deferred();
+       var nodes = [];
+       var treeID;
+       for(var id in features){
+
+           var box = turf.bbox(features[id]);
+           features[id].minX = box[0];
+           features[id].minY = box[1];
+           features[id].maxX = box[2];
+           features[id].maxY = box[3];
+           features[id].properties["centerLog"] = (features[id].maxX + features[id].minX) / 2;
+           features[id].properties["centerLat"] = (features[id].maxY + features[id].minY) / 2;
+           treeID = box[0]+"/"+box[1]+"/"+box[2]+"/"+box[3];
+           if( insertedTreeIDs.has(treeID) == false){
+                   nodes.push(features[id]);
+                   insertedTreeIDs.add(treeID);
+                   cacheSize = cacheSize + 1;
+           }
+
+       }
+
+       if(cacheSize >= cacheThreshold){
+
+             DeleteTarget = cacheSize - cacheThreshold;
```

Comment:
This is the place where we know our cache has reached the full size and we need to evict polygons to insert them